### ISSUE TITLE
[Question / Feature Suggestion] Add infinite loop detection

### ISSUE THREAD

**Issue by slisaasquatch** [IBM/JSONata4Java#23]
jsonata-js allows the user to do this to detect infinite loops:
![](https://i.imgur.com/AWlMxEH.png)

Is it possible to do something like it with this library?

**Comment by wnm3** [comment]
Thanks, I'll take a look and see what is possible.

**Comment by wnm3** [comment]
I will prototype this in the same code used to showcase function declarations and then move to add this as a general feature so people can add these limits along with their expression when executing JSONata for Java.

**Comment by wnm3** [comment]
I have begun the check in process to post a new version, 1.2 of JSONata4Java to Maven Central. It includes a bug fix for issue #24 as well as tooling you've requested.

I created a new TesterTimeBox utility that will prompt you for a timeoutMS (max milliseconds of execution time before throwing an exception) and a maxDepth (the number of recursive calls allowed before throwing an exception).

I also added a new method to the Expressions class allowing you to pass the timeoutMS and maxDepth when evaluating your expression (see the [TesterTimeBox](https://github.com/IBM/JSONata4Java/blob/master/src/main/java/com/api/jsonata4java/TesterTimeBox.java) code to see how it was called).
```
   /**
    * Evaluate the stored expression against the supplied event and application
    * interface data.
    * 
    * @param rootContext bound to root context ($$ and paths that don't start with
    *                    $event, $state or $instance) when evaluating expressions.
    *                    May be null.
    * @param timeoutMS milliseconds allowed for the evaluation to occur. If it takes longer 
    *                    an exception is thrown. Must be positive number or exception is thrown.
    * @param maxDepth the maximum call stack depth allowed before an exception is thrown. Must 
    *                    be a positive number or an exception is thrown.
    * @return the JsonNode resulting from the expression evaluation against the rootContext
    * @throws EvaluateException If the given device event is invalid.
    */
	public JsonNode evaluate(JsonNode rootContext, long timeoutMS, int maxDepth) throws EvaluateException
```

Please close and comment if this provides the solution to your request. Thank you.

**Comment by slisaasquatch** [comment]
Thanks a lot! I'll close the ticket.
